File: src/original-c#-camera/PPGSignalsCollector.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading;
using CamPPG;

namespace RivaDigitalApp
{
    public abstract class PPGSignalsCollector : IPPGSignalsCollector
    {
        protected readonly CamPPG.Video2PPGConverter _converter;

        private readonly ConcurrentBag<PpgSignal> signals = new ConcurrentBag<PpgSignal>();

        private int qualityWarnings;

        protected PPGSignalsCollector (int width, int height, int phoneModel)
        {
            _converter = new CamPPG.Video2PPGConverter();
            _converter.InitAlgo(width, height, phoneModel);
        }

        public static IPPGSignalsCollector CreateDroidCollector(int width, int height)
        {
            return new DroidSignalsCollector(width, height);
        }

        public static IPPGSignalsCollector CreateiOsCollector(int width, int height)
        {
            return new iOSSignalsCollector(width, height);
        }

        public IEnumerable<PpgSignal> Data => signals;

        public int SignalsCount => signals.Count;

        public int QualityWarnings => qualityWarnings;

        public void ProcessFrame(Frame frame)
        {
            ConvertFrame(frame, out var ppgStruct);
            if (ppgStruct.QualityWarning)
            {
                Interlocked.Increment(ref qualityWarnings);
            }

            signals.Add(new PpgSignal() { Timestamp = ppgStruct.timestamp, Signals = ppgStruct.Signals, QualityWarning = ppgStruct.QualityWarning });
        }

        //TODO: This functionality should be part of IPPGSignalsProcessor
        protected abstract void ConvertFrame(Frame frame, out Video2PPGConverter.PpgStruct ppgStruct);

        class iOSSignalsCollector : PPGSignalsCollector
        {
            public iOSSignalsCollector(int width, int height) : base(width, height, 2)
            {
            }

            protected override void ConvertFrame(Frame frame, out Video2PPGConverter.PpgStruct ppgStruct)
            {
                _converter.ConvertFrame2PPGiOS(frame.Timestamp, frame.Y, frame.UV, out ppgStruct);
            }
        }

        class DroidSignalsCollector : PPGSignalsCollector
        {
            public DroidSignalsCollector(int width, int height) : base(width, height, 1)
            {
            }

            protected override void ConvertFrame(Frame frame, out Video2PPGConverter.PpgStruct ppgStruct)
            {
                _converter.ConvertFrame2PPGAndroid(frame.Timestamp, frame.Y, frame.U, frame.V, out ppgStruct);
            }
        }
    }
}

File: src/swift-camera/PPGSignalsCollector.swift
import Foundation

// Swift equivalent of PPGSignalsCollector.cs
// Note: This assumes the existence of a Video2PPGConverter Swift wrapper for the CamPPG library

class PPGSignalsCollector: IPPGSignalsCollector {
    protected let _converter: Video2PPGConverter
    private var signals: [PpgSignal] = []
    private var qualityWarnings: Int = 0
    private let signalsQueue = DispatchQueue(label: "signals.queue", attributes: .concurrent)
    
    var signalsCount: Int {
        return signalsQueue.sync { signals.count }
    }
    
    var qualityWarnings: Int {
        return signalsQueue.sync { self.qualityWarnings }
    }
    
    var data: [PpgSignal] {
        return signalsQueue.sync { signals }
    }
    
    init(width: Int, height: Int, phoneModel: Int) {
        _converter = Video2PPGConverter()
        _converter.initAlgo(width: width, height: height, phoneModel: phoneModel)
    }
    
    static func createDroidCollector(width: Int, height: Int) -> IPPGSignalsCollector {
        return DroidSignalsCollector(width: width, height: height)
    }
    
    static func createiOsCollector(width: Int, height: Int) -> IPPGSignalsCollector {
        return iOSSignalsCollector(width: width, height: height)
    }
    
    func processFrame(_ frame: Frame) {
        let ppgStruct = convertFrame(frame)
        
        if ppgStruct.qualityWarning {
            signalsQueue.async(flags: .barrier) {
                self.qualityWarnings += 1
            }
        }
        
        let ppgSignal = PpgSignal(
            timestamp: ppgStruct.timestamp,
            signals: ppgStruct.signals,
            qualityWarning: ppgStruct.qualityWarning
        )
        
        signalsQueue.async(flags: .barrier) {
            self.signals.append(ppgSignal)
        }
    }
    
    // TODO: This functionality should be part of IPPGSignalsProcessor
    func convertFrame(_ frame: Frame) -> PpgStruct {
        fatalError("convertFrame must be overridden by subclass")
    }
}

// iOS-specific signals collector
class iOSSignalsCollector: PPGSignalsCollector {
    init(width: Int, height: Int) {
        super.init(width: width, height: height, phoneModel: 2)
    }
    
    override func convertFrame(_ frame: Frame) -> PpgStruct {
        return _converter.convertFrame2PPGiOS(
            timestamp: frame.timestamp,
            Y: frame.Y,
            UV: frame.UV
        )
    }
}

// Android-specific signals collector
class DroidSignalsCollector: PPGSignalsCollector {
    init(width: Int, height: Int) {
        super.init(width: width, height: height, phoneModel: 1)
    }
    
    override func convertFrame(_ frame: Frame) -> PpgStruct {
        return _converter.convertFrame2PPGAndroid(
            timestamp: frame.timestamp,
            Y: frame.Y,
            U: frame.U,
            V: frame.V
        )
    }
}

// Placeholder for the Video2PPGConverter and PpgStruct
// These would need to be implemented as Swift wrappers around the CamPPG library
class Video2PPGConverter {
    func initAlgo(width: Int, height: Int, phoneModel: Int) {
        // Implementation would call into CamPPG library
    }
    
    func convertFrame2PPGiOS(timestamp: UInt64, Y: Data, UV: Data) -> PpgStruct {
        // Implementation would call into CamPPG library
        return PpgStruct(timestamp: timestamp, signals: [], qualityWarning: false)
    }
    
    func convertFrame2PPGAndroid(timestamp: UInt64, Y: Data, U: Data, V: Data) -> PpgStruct {
        // Implementation would call into CamPPG library
        return PpgStruct(timestamp: timestamp, signals: [], qualityWarning: false)
    }
}

struct PpgStruct {
    let timestamp: UInt64
    let signals: [Double]
    let qualityWarning: Bool
}
